src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,407 @@
+
+	private int idleBetweenPolls;
+
+	private int consumerBackoff;
+
+	private long checkpointsInterval = 60_000L;
+
+	public KclMessageDrivenChannelAdapter(String streams, Executor executor) {
+		this(streams, executor, KinesisAsyncClient.builder().build(),
+				CloudWatchAsyncClient.builder().build(), DynamoDbAsyncClient.builder().build());
+	}
+
+	public KclMessageDrivenChannelAdapter(String streams, Executor executor, Region region) {
+		this(streams, executor, KinesisAsyncClient.builder().region(region).build(),
+				CloudWatchAsyncClient.builder().region(region).build(), DynamoDbAsyncClient.builder().region(region).build());
+	}
+
+	public KclMessageDrivenChannelAdapter(String stream, Executor executor,
+			KinesisAsyncClient kinesisClient, CloudWatchAsyncClient cloudWatchClient, DynamoDbAsyncClient dynamoDBClient) {
+		Assert.notNull(stream, "'stream' must not be null.");
+		Assert.notNull(executor, "'executor' must not be null.");
+		Assert.notNull(kinesisClient, "'kinesisClient' must not be null.");
+		Assert.notNull(cloudWatchClient, "'cloudWatchClient' must not be null.");
+		Assert.notNull(dynamoDBClient, "'dynamoDBClient' must not be null.");
+		this.stream = stream;
+		this.executor = executor;
+		this.kinesisClient = kinesisClient;
+		this.cloudWatchClient = cloudWatchClient;
+		this.dynamoDBClient = dynamoDBClient;
+	}
+
+	public void setConsumerGroup(String consumerGroup) {
+		Assert.hasText(consumerGroup, "'consumerGroup' must not be empty");
+		this.consumerGroup = consumerGroup;
+	}
+
+	/**
+	 * Specify an {@link InboundMessageMapper} to extract message headers embedded
+	 * into the record data.
+	 *
+	 * @param embeddedHeadersMapper the {@link InboundMessageMapper} to use.
+	 * @since 2.0
+	 */
+	public void setEmbeddedHeadersMapper(InboundMessageMapper<byte[]> embeddedHeadersMapper) {
+		this.embeddedHeadersMapper = embeddedHeadersMapper;
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+
+		String workerId = UUID.randomUUID().toString();
+		RecordProcessorFactory recordProcessorFactory = new RecordProcessorFactory();
+
+		ConfigsBuilder configsBuilder = new ConfigsBuilder(this.stream, this.consumerGroup,
+				this.kinesisClient, this.dynamoDBClient, this.cloudWatchClient, workerId, recordProcessorFactory);
+		configsBuilder.retrievalConfig().initialPositionInStreamExtended(this.streamInitialSequence);
+		configsBuilder.retrievalConfig().listShardsBackoffTimeInMillis(this.consumerBackoff);
+		configsBuilder.coordinatorConfig().parentShardPollIntervalMillis(this.idleBetweenPolls);
+
+		this.scheduler = new Scheduler(configsBuilder.checkpointConfig(),
+			configsBuilder.coordinatorConfig(),
+			configsBuilder.leaseManagementConfig(),
+			configsBuilder.lifecycleConfig(),
+			configsBuilder.metricsConfig(),
+			configsBuilder.processorConfig(),
+			configsBuilder.retrievalConfig());
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+		this.executor.execute(() -> this.scheduler.run());
```

Comment:
Looks like the `Scheduler implements Runnable`, no reason to wrap it into another `Runnable`:
```suggestion
		this.executor.execute(this.scheduler);
```